packages/js/client-config-builder/src/ClientConfigBuilder.ts

```diff
@@ -4,11 +4,11 @@ import { BaseClientConfigBuilder } from "./BaseClientConfigBuilder";
 import { CoreClientConfig, Uri, IUriResolver } from "@polywrap/core-js";
 import {
   IWrapperCache,
-  PackageToWrapperCacheResolver,
   RecursiveResolver,
   StaticResolver,
   WrapperCache,
 } from "@polywrap/uri-resolvers-js";
+import { EcoCacheResolver } from "wraplib";
```

Comment:
Why don't we move the resolver to uri-resolver-js?

Reply:
Can we also do the same for the other utility functions and move it to its relevant package in toolchain?

Reply:
I think `uri-resolvers-js` is for core/fundamental resolvers. I imagine in the future there will be many resolvers developed by a large Polywrap community. If it makes sense to put the `EcoCacheResolver` in `uri-resolvers-js`, I'm happy to do it.

I created the wraplib package for my own exploration and enjoyment. I had some ideas for things I wanted to try out. It was motivating when I was exhausted. 

I can move the wraplib repo to the Polywrap organization so long as I can continue to freely develop in it as I please, but I'm not going to move everything out of it and delete it.

Reply:
I added `EcoCacheResolver` to `uri-resolvers-js` and removed `wraplib` as a dependency. Instead of having two similar resolvers, would it make sense to instead update `PackageToWrapperCacheResolver` with the new features and refactor? Or should they remain distinct resolvers?
 
I would really like your input on this @nerfZael

Reply:
I thought we are going to improve `PackageToWrapperCacheResolver` with `EcoCacheResolver`, Isn't that the case?

Reply:
I didn't know that's what you were requesting.

I pushed a commit that removes `EcoCacheResolver` completely and instead modifies `PackageToWrapperCacheResolver`.

Reply:
I don't think this should be part of PackageToWrapperCacheResolver.
It's better to separate it into its own resolver, so that it can be used with any other resolver (just like PackageToWrapperCacheResolver or RecursiveResolver can)

Reply:
It can be a distinct resolver, but it would be a drop-in alternative for the `PackageToWrapperCacheResolver`. Is that okay?

It needs to be able to check if a URI is cached. So it either has to be integrated with the `PackageToWrapperCacheResolver` or take a `PackageToWrapperCacheResolver` as the resolver argument in its constructor/factory method.

What do you think?

Reply:
I'm thinking something along the lines of a distinct resolver that has it's own cache and takes any IUriResolver as argument (In the default config that would be the PackageToWrapperCacheResolver).
The cache could be Uri -> UriResolutionResult (Result<UriPackageOrWrapper, Error>)

Reply:
Actually, in the default config, PackageToWrapperCacheResolver could take this new resolver as argument. That might be better

Reply:
Not a fan of EcoResolver name though, maybe call it RequestSynchronizationResolver? I dunno :D

Reply:
> I'm thinking something along the lines of a distinct resolver that has it's own cache and takes any IUriResolver as argument (In the default config that would be the PackageToWrapperCacheResolver). The cache could be Uri -> UriResolutionResult (Result<UriPackageOrWrapper, Error>)

This is an interesting idea. I'll work on this.

Reply:
I don't think it makes

> Actually, in the default config, PackageToWrapperCacheResolver could take this new resolver as argument. That might be better

The idea is that the lock is in place until the wrapper is cached, so I think it makes more sense for it to take the `PackageToWrapperCacheResolver` as an argument.

Reply:
> I'm thinking something along the lines of a distinct resolver that has it's own cache and takes any IUriResolver as argument (In the default config that would be the PackageToWrapperCacheResolver). The cache could be Uri -> UriResolutionResult (Result<UriPackageOrWrapper, Error>)

I'm implementing this and I am a little worried that having a separate cache will effectively double the amount of data cached and eat up memory. Do you think that's an issue?

Reply:
No, but I also messed up when talking about the type of cache.
It should not be Uri -> Result<UriPackageOrWrapper, Error>
It should be Uri -> Promise<Result<UriPackageOrWrapper, Error>>
And ofc, the items are temporary and should be deleted after promise resolution.

Reply:
Okay, gotcha

Reply:
I thought I understood what you were saying but now I don't think I do.

Let's say there are N invocations waiting to obtain the mutex and 1 resolving. The resolving invocation finishes resolution and returns a `Result<UriPackageOrWrapper, TError>`. The lock is destroyed. The remaining 9 need to return the `Result<UriPackageOrWrapper, TError>`. They should not be forwarded to the next resolver.

How do I know when the `Result<UriPackageOrWrapper, TError>` is safe to delete?

In other words, how do I know that the waiting "threads" have finished returning the `Result` and the wrapper has been successfully cached by the `PackageToWrapperCacheResolver`?

I don't think storing a `Promise` is helpful because we need to see if the result is an error before we can destroy the lock.

Reply:
I could set a timeout to delete the `Result<UriPackageOrWrapper, TError>` but that isn't reliable.

Also, it really makes more sense to store the `UriPackageOrWrapper`. If a result is an error then we can't cache anything and we don't want to destroy the mutex yet.